spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSourceTests.java

```diff
@@ -0,0 +1,258 @@
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(1);
+		pubSubMessageSource.setPayloadType(String.class);
+		pubSubMessageSource.setAckMode(AckMode.AUTO);
+		MessageSourcePollingTemplate poller = new MessageSourcePollingTemplate(pubSubMessageSource);
+		poller.poll((message) -> {
+			assertThat(message).isNotNull();
+
+			assertThat(message.getPayload()).isEqualTo("msg1");
+			assertThat(message.getHeaders().get(GcpPubSubHeaders.ORIGINAL_MESSAGE)).isEqualTo(this.msg1);
+		});
+
+		verify(this.msg1).ack();
+	}
+
+	@Test
+	public void doReceive_autoAckModeAcksAndAddsOriginalMessageHeader() {
+
+		PubSubMessageSource pubSubMessageSource = new PubSubMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(1);
+		pubSubMessageSource.setPayloadType(String.class);
+		pubSubMessageSource.setAckMode(AckMode.AUTO_ACK);
+		MessageSourcePollingTemplate poller = new MessageSourcePollingTemplate(pubSubMessageSource);
+		poller.poll((message) -> {
+			assertThat(message).isNotNull();
+
+			assertThat(message.getPayload()).isEqualTo("msg1");
+			assertThat(message.getHeaders().get(GcpPubSubHeaders.ORIGINAL_MESSAGE)).isEqualTo(this.msg1);
+		});
+
+		verify(this.msg1).ack();
+	}
+
+
+	@Test
+	public void doReceive_autoModeNacksAutomatically() {
+
+		PubSubMessageSource pubSubMessageSource = new PubSubMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(1);
+		pubSubMessageSource.setPayloadType(String.class);
+		pubSubMessageSource.setAckMode(AckMode.AUTO);
+		MessageSourcePollingTemplate poller = new MessageSourcePollingTemplate(pubSubMessageSource);
+		assertThatThrownBy(() -> {
+			poller.poll((message) -> {
+				throw new RuntimeException("Nope.");
+			});
+		}).isInstanceOf(MessageHandlingException.class).hasMessageContaining("Nope.");
+
+		verify(this.msg1).nack();
+	}
+
+	@Test
+	public void doReceive_autoAckModeDoesNotNackAutomatically() {
+
+		PubSubMessageSource pubSubMessageSource = new PubSubMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(1);
+		pubSubMessageSource.setPayloadType(String.class);
+		pubSubMessageSource.setAckMode(AckMode.AUTO_ACK);
+		MessageSourcePollingTemplate poller = new MessageSourcePollingTemplate(pubSubMessageSource);
+		assertThatThrownBy(() -> {
+			poller.poll((message) -> {
+				throw new RuntimeException("Nope.");
+			});
+		}).isInstanceOf(MessageHandlingException.class).hasMessageContaining("Nope.");
+
+		verify(this.msg1, times(0)).nack();
+	}
+
+	@Test
+	public void blockOnPullSetsReturnImmediatelyToFalse() {
+
+		PubSubMessageSource pubSubMessageSource = new PubSubMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(1);
+		pubSubMessageSource.setBlockOnPull();
+
+		Object message = pubSubMessageSource.doReceive(1);
```

Comment:
I guess we need to *assert* this for `null` at least.

And also, please, add `@SuppressWarning("unchecked")` for all the method where we have such a warning, if you can't fix generics, of course...